### ISSUE TITLE
Suggestion for Calico eBPF: Introduction of Direct Routing Mode

### ISSUE THREAD

**Issue by pf2rich** [projectcalico/calico#8620]
In our current use of Calico's eBPF mode, we have noticed a significant drop in maximum bandwidth—over 50% less compared to when in IPVS mode—for traffic entering the Kubernetes cluster from outside. This reduction occurs as traffic moves via VXLAN from the ingress node to the workload node. We find this decrease in performance to be unreasonable. Is there any plan in the community to implement a Direct Routing (DR) mode, similar to Cilium's DR mode, to route traffic directly to the target pod instead of encapsulating packets in VXLAN?

**Comment by tomastigera** [comment]
@pf2rich I am not familiar with cilium's direct routing. Would you have any pointers or some explanation of it?

50% reduction is quite a bit and certainly worth investigation? Do you observe some increased load on only certain (one) CPU on the communicating nodes? I would be searching for some perf degradation due to locality of the vxlan nodeport traffic.

**Comment by pf2rich** [comment]
@tomastigera 
Based on my understanding of Cilium, when accessing services through NodePort, as traffic reaches node A, Cilium modifies the packet's destination IP address/port to the Pod's IP address/port, and records the original IP and port information in the IP option field. Then, it is routed to the Pod through layer 3 routing. The packets returned by the Pod are subject to SNAT operations through eBPF.

**Comment by tomastigera** [comment]
@pf2rich do you have bonded NICs on your nodes by any chance?

**Comment by pf2rich** [comment]
Yes, we have configured the two physical network cards as bond4.

**Comment by tomastigera** [comment]
It seems that the 50% drop is due to the fact that only one of the devices is used. And since since you saturate the link in/out you can only use half the bandwidth. That should be the same for ebpf or not. However, with bond, you _may_ and possibly _should_ be able to use half the bandwidth of the bond.

(1) in ebpf mode you would have by default the ebpf program on each of the phys devices. It is arguable whether that is a good thing and whether it should be on the bond itself. If you use strict rpf, it must be on the bond.
(2) even when the program is hooked on the bond, it seems that only one of the phys devices is used for both in/out traffic
(3) iptables/ipvs do not have that problem with calico afaict as routing traffic back (forwarded to another node) is done much later
(4) the vxlan tunnel itself does not seem to be the issue for the 50% drop
(5) If vxlan overlay is used for pods (which it often is) forwarding external traffic to a service via the vxlan tunnel is actually faster than using the overlay.

All this said, I am looking into why ebpf does not use both phys devices underlying the bond.

**Comment by tomastigera** [comment]
@pf2rich what is your bond mode?

**Comment by esimest** [comment]
> @pf2rich what is your bond mode? My mode is 802.3ad layer3+4 so the vxlan tunnel between two nodes is always pinned to the same phys dev.

cat /sys/class/net/bond0/bonding/mode
802.3ad 4

**Comment by esimest** [comment]
@tomastigera I test this commit in my environment. Traffic using ebpf mode is now the same as traffic using ipvs mode

**Comment by tomastigera** [comment]
@esimest thanks for testing it! Much appreciated!

**Comment by tomastigera** [comment]
Although the fix for the issue that triggered this suggestion is merged, I will keep this it open for the sake of the suggestion.
